core/network/src/protocol/sync.rs

```diff
@@ -1258,3 +1270,32 @@ fn peer_block_request<B: BlockT>(
 		None
 	}
 }
+
+/// Get pending explicit sync request for a peer.
+fn explicit_sync_request<B: BlockT>(
+	id: &PeerId,
+	requests: &HashMap<B::Hash, SyncRequest<B>>,
+	best_num: NumberFor<B>,
+	finalized: &B::Hash,
+	attributs: &message::BlockAttributes,
+) -> Option<(B::Hash, BlockRequest<B>)>
+{
+	for (hash, r) in requests {
+		if !r.peers.contains(id) {
+			continue
+		}
+		if r.number <= best_num {
```

Comment:
questionable for the same reason as before - the orphan chain may be longer than the peer's "best" chain because of the fork-choice rule.

Reply:
The other case (`r.number > best_num`) is handled in a different code path in `peer_block_request`.

Reply:
This is confusing. Why is the logic so scattered?

Reply:
What I mean is that there's a lot of implicit "but actually" in the interface of this function. 

It's an explicit sync request...but actually only returns explicit requests for chains shorter than the best.

So the peer doesn't download that chain...but actually it does, it's covered in another function in another file.

Reply:
I'd recommend that the function return a special type when there is implicit context that needs to be passed from the execution of this function into the execution of another, where the reader of the code can clearly see that the handling of some case has been deferred beyond the execution of the function.

Reply:
This is fundamentally caused by the same issue described here: https://github.com/paritytech/substrate/pull/3633#discussion_r326888400
To be refactored.